Implement `readline(s)` for all stream-seek classes

t-reents opened the pull request:
Adding a `readline` method to `PackedObjectReader` to address #174. This is important as it causes errors in `aiida-workgraph` and `aiida-pythonjob`.

According to the test and some manual comparison, it seems that this is working fine. Would be great if we could quickly iterate and get this in, as @superstar54 was also keen in fixing this for the `WorkGraph/PythonJob`

t-reents committed:
Add `readline` method to `PackedObjectReader`

This commit adds a `readline` method to the `PackedObjectReader`, which
was necessary to make it comatible with pickle. The current version
causes problems in `aiida-workgraph` and `aiida-pythonjob`, see #174.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

t-reents commented:
Pinging @agoscinski @GeigerJ2 @khsrali for a review, thanks!

t-reents committed:
Change naming

agoscinski commented on disk_objectstore/utils.py:
```diff
@@ -615,6 +615,17 @@ def read(self, size: int = -1) -> bytes:
+            return b''
+
+        readline_size = remaining if (size is None or size < 0) else min(size, remaining)
+        line = self._fhandle.readline(readline_size)
```
So this reads only up until some terminator, at least for BinaryIO this is `b'\n'`, even though it can be changed for some file handler, the `readline_size` is just a maximum value up until it reads. I am not sure if this is the intended behavior, since my understanding of the `PackedObjectReader` is that it contains already "one line". It seems like you just make reading slower since you randomly chunk your data whenever the you find a `b'\n'` which has no meaning for arbitrary packed objects (unlike your text example in the tests). What is more meaningful is to have some maximum number to limit memory consumption.

t-reents replied:
Thanks Alex!
I generally agree and was also thinking if `b'\n'` really makes sense, or something "fixed". In the end, I followed the "new-line" approach, based on what I saw for the `readline` method of the `BytesIO` / `BufferedReader` that would be used when the repository isn't packed https://github.com/python/cpython/blob/main/Lib/_pyio.py#L509-L518. Moreover,  think that I was mostly biased by the pickle application of that. I just had a quick look there and it seems that `pickle` is also most relying on `b'\n'` in the readline method. 
But again, I agree that this might not be the most reasonable approach for such a packed file.

agoscinski reviewed (commented):
Explaining mypy error: So the `PackedObjectReader` accepts fhandlers of type `StreamSeekBytesType` which is defined as 
```
StreamSeekBytesType = Union[
    BinaryIO,
    'PackedObjectReader',
    'CallbackStreamWrapper',
    'ZlibLikeBaseStreamDecompresser',
]
```
Since PackedObjectReader readline method uses the fhandler's readline method, the type checker correctly gives you the error that you need to implement this function for these file handlers as well. If you consider not to implement the methods in the other classe, you at least need to do an isinstance check and error out in case it is a type that does not support readline.

GeigerJ2 committed:
add code from meeting

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

GeigerJ2 committed:
wip: readline(s) for all stream seek classes, tests, and benchmarks

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

GeigerJ2 committed:
remove leftover files

GeigerJ2 committed:
fix types

GeigerJ2 committed:
Merge branch 'main' into fix/PackedObjectReader_readline

GeigerJ2 committed:
use protocols for types

GeigerJ2 committed:
Merge branch 'main' into fix/PackedObjectReader_readline

GeigerJ2 committed:
Revert "remove leftover files"

This reverts commit 24b7e31d081b33c9e07897528c719783266995ae.

GeigerJ2 committed:
wip

GeigerJ2 committed:
Merge remote-tracking branch 'upstream/main' into fix/PackedObjectReader_readline

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

GeigerJ2 committed:
move test-speed.ipynb to shared scripts location

GeigerJ2 committed:
Merge branch 'main' into fix/PackedObjectReader_readline

GeigerJ2 committed:
wip

GeigerJ2 committed:
⏪ `utils`: revert stream-guard dedup to a follow-up

Consolidating the repeated guards in `LazyLooseStream` and
`ZlibLikeBaseStreamDecompresser` (the `_require_stream` /
`_require_uncompressed_stream` accessors that centralised the "I/O
operation on closed file" and "uncompressed stream is None" checks) is a
pure refactor of pre-existing code, unrelated to adding
`readline`/`readlines`.

Reverting it here keeps this PR scoped to the readline feature and the
decompression regression fix it introduced. The guard deduplication will
be raised as a separate follow-up issue/PR.

GeigerJ2 committed:
Revert "⏪ `utils`: revert stream-guard dedup to a follow-up"

This reverts commit f9616e5326e0d4e0e9fc71c3b058a9121e8657a6.

GeigerJ2 committed:
♻️ `utils`: dedupe `readlines` via a shared helper

This PR adds `readlines` to `PackedObjectReader`,
`CallbackStreamWrapper`, and `ZlibLikeBaseStreamDecompresser` with an
identical "loop over readline until EOF / hint" body. Rather than ship
the same loop three times, collect it into a module-level
`_collect_readlines(readline, hint)` helper (mirroring
`io.IOBase.readlines`) that each class calls with its own `readline`.

Behaviour is unchanged; the helper preserves the `hint`-bytes cutoff and
the read-all default. Pre-existing stream-wrapper duplication (the
read-only boilerplate, `read()`, chunk-size constant) is left for a
separate follow-up.

GeigerJ2 committed:
👌 `container`: drop pack-stream type-ignores

`add_streamed_objects_to_pack` annotated `stream_context_manager` as
`AbstractContextManager[StreamSeekBytesType | BinaryIO]` but then
suppressed three mypy errors with `# type: ignore`. The root cause was
`working_stream_list = list(stream_list[::-1])`: mypy joins the
`list[StreamSeekBytesType] | list[LazyOpener]` parameter down to
`list[object]`, so `next_stream` was typed `object` and nothing assigned
from it type-checked.

Annotating `working_stream_list` with the element union keeps
`next_stream` as `StreamSeekBytesType | LazyOpener`, both of which mypy
accepts directly as the context-manager type (they are structurally
context managers), so the `[assignment]` ignore is no longer needed. The
`[arg-type]` ignore on the `nullcontext` branch was already redundant.
With the context manager correctly typed, the `stream` bound by the
`with` block is already `StreamSeekBytesType | BinaryIO`, so the
`stream_typed` alias and its ignore are removed in favour of using
`stream` directly.

GeigerJ2 committed:
🧪 `test_utils`: cover new readline/readlines paths

Codecov flagged the new stream code as under-covered. Add real tests (no
mocks) for the branches the existing suite never exercised:

- `readline(size)` and `readlines(hint)` across all stream classes,
  parametrized over `BytesIO`, `PackedObjectReader`,
  `CallbackStreamWrapper` and `ZlibStreamDecompresser` (the hint cases
  cover both the hint-reached and end-of-stream paths).
- `CallbackStreamWrapper.readline` driving the callback, asserted via
  the recording `callback_instance` fixture (every byte accounted for
  once).
- `LazyLooseStream.readline`/`readlines` directly, and the compressed
  stream delegating to its loose copy (with `peek` returning empty
  there) after a backward seek materialises it.

Also exclude bare `...` Protocol method bodies from coverage in
`.coveragerc`; they are non-executable stubs. Together these bring patch
coverage of the new `utils.py`/`container.py` lines to 100%.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

GeigerJ2 committed:
⚡️ `ZlibStreamDecompresser`: O(n) `readline`

`readline` over a compressed object was O(buffer * lines): every `read`
re-sliced the internal buffer with `buffer[size:]`, copying the ~512 KB
remainder once per line, and `peek` copied it again. On large,
line-structured objects this was super-linear and effectively unusable
(a 4 MB text object took ~340 ms vs ~10 ms for a full `read`).

Consume the decompressed buffer through a `_buffer_pos` offset instead
of re-slicing it (dropping the consumed prefix only on refill), and
locate newlines in place with `bytes.find(b'\n', pos)`. Both copies are
gone, so a line costs time proportional to its length and `readline` is
linear in object size: ~5-8x faster on multi-MB text, and the
readline/read ratio no longer grows with size.

`peek` is removed. It was added in this PR only to drive the old
`readline`, was never released, and was inconsistent (none of the other
stream wrappers expose it, nor does the stream Protocol); the in-place
search makes it unnecessary. `_buffer_pos` is reset in `__init__` and on
the seek-to-zero fast path.

Also add reproducible read-vs-readline benchmarks in
`tests/test_benchmark.py`, across object sizes and shapes (few-line to 4
MB, text/binary/blob, both packed paths), to lock in a speed baseline.
Behaviour is otherwise unchanged and covered by the existing
decompresser/readline/seek tests.

GeigerJ2 committed:
🧪 `test_utils`: readline over many short lines

The offset-based buffer consumption in `ZlibStreamDecompresser.readline`
is exercised most by many short lines crossing the 512 KB buffer
boundary: each line advances the read offset within the buffer, and the
buffer is refilled between lines. That path was only asserted in the
readline benchmarks, which CI runs with `--benchmark-skip`.

Add a fast unit test that reads ~1 MB of distinct numbered lines back
through a compressed stream and checks they reconstruct exactly, so the
path is covered in the normal test run and a dropped, duplicated or
reordered line would fail.

GeigerJ2 committed:
⚡️ `PackedObjectReader`: faster `readline`

`readline` ended every call with `_update_pos`, which re-derives the
position with a per-line `self._fhandle.tell()` round-trip plus a method
frame. Since `readline` consumes exactly `len(line)` bytes (bounded by
`min(size, remaining)`, so it cannot cross the object boundary), advance
`self._pos` by `len(line)` directly instead.

~2.6x faster for line-by-line reads of large uncompressed packed objects
(27.7 -> 9.7 ms over a 4 MB object of 65-byte lines); negligible
elsewhere. Behaviour is unchanged: the `min(size, remaining)` cap still
guarantees the boundary that `_update_pos` used to assert.

GeigerJ2 committed:
♻️ `test_util`: consolidate readline tests, relocate helper
